src/reflect/scala/reflect/internal/tpe/TypeComparers.scala

```diff
@@ -365,9 +365,23 @@ trait TypeComparers {
 
   // @assume tp1.isHigherKinded || tp2.isHigherKinded
   def isHKSubType(tp1: Type, tp2: Type, depth: Depth): Boolean = {
-    def isSub(ntp1: Type, ntp2: Type) = (ntp1.withoutAnnotations, ntp2.withoutAnnotations) match {
+    def isSub(tp1: Type, tp2: Type) = (tp1.withoutAnnotations, tp2.withoutAnnotations) match {
       case (TypeRef(_, AnyClass, _), _)                                     => false                    // avoid some warnings when Nothing/Any are on the other side
       case (_, TypeRef(_, NothingClass, _))                                 => false
+      case (tv1 @ TypeVar(_, _), tv2 @ TypeVar(_, _)) if settings.YhkTypevarUnification =>
+        (tv2.params corresponds tv1.params)(methodHigherOrderTypeParamsSubVariance) &&
+        { tv1.addHiBound(tv2); tv2.addLoBound(tv1); true }
+      case (tp1, tv2 @ TypeVar(_, _)) if settings.YhkTypevarUnification =>
+        val ntp1 = tp1.normalize
+        (tv2.params corresponds ntp1.typeParams)(methodHigherOrderTypeParamsSubVariance) &&
+        { tv2.addLoBound(ntp1); true }
```

Comment:
The existing code in `TypeVar.registerBound` that responsible for unifying applied hk type vars walks up the base type sequence to find an arity match. What do we know about `tv2.params.length == ntp1.typeParams.length` by virtue of making it to this line of code?

Reply:
We know `tv2.params.length == ntp1.typeParams.length` thanks to a successful `corresponds` on the previous line.